### PR TITLE
Add rest API for take/free instance

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

#1896

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This change adds take/free instance rest API endpoint and JSON parser.

### Tests

- [X] The following tests are written for this issue:

Test will be added when take/freeInstance is implemented in clusterMaintenanceService.java 

- The following is the result of the "mvn test" command on the appropriate module:

```
[INFO] Tests run: 179, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 287.922 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 179, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.6:report (generate-code-coverage-report) @ helix-rest ---
[INFO] Loading execution data file /Users/xialu/Documents/WorkSpace/helix/helix-rest/target/jacoco.exec
[INFO] Analyzed bundle 'Apache Helix :: Restful Interface' with 85 classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:52 min
[INFO] Finished at: 2021-11-30T11:26:48-08:00
[INFO] ------------------------------------------------------------------------
```

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java [hunk 1]
```diff
@@ -412,17 +411,36 @@ private StoppableCheck performCustomInstanceCheck(String clusterId, String insta
...
+
     JsonNode jsonNode = OBJECT_MAPPER.readTree(jsonContent);
     // parsing the inputs as string key value pairs
     jsonNode.fields().forEachRemaining(kv -> result.put(kv.getKey(), kv.getValue().asText()));
     return result;
   }
 
+  public static List<String> getListFromJsonPayload(JsonNode jsonContent) throws IOException {
+    if (jsonContent == null) {
+      return new ArrayList<>();
```
Comment (narendly) [inline helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:428]: `Collections.emptyList()` might be better here?

Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java]: TFTR. Updated.

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java [hunk 2]
```diff
@@ -423,6 +422,23 @@ private StoppableCheck performCustomInstanceCheck(String clusterId, String insta
...
+    }
+    return OBJECT_MAPPER.convertValue(jsonContent, List.class);
+  }
+
+  public static Map<String, String> getMapFromJsonPayload(JsonNode jsonContent) throws IOException {
+    Map<String, String> result = new HashMap<>();
+    if (jsonContent == null) {
+      return result;
+    }
+    result = OBJECT_MAPPER.convertValue(jsonContent, new TypeReference<Map<String, String>>() {
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:437]: Would this throw exception?

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java [hunk 3]
```diff
@@ -185,6 +200,137 @@ public Response isInstanceStoppable(
...
+
+    try {
+      JsonNode node = null;
+      if (jsonContent.length() != 0) {
+        node = OBJECT_MAPPER.readTree(jsonContent);
+      }
+      if (node == null) {
+        return badRequest("Invalid input for content : " + jsonContent);
+      }
+      MaintenanceOpInputFields inputFields = readMaintenanceInputFromJson(node);
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java:275]: This is dup code with previous function. Let's wrap it into readMaintenanceInputFromJson function?

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java [hunk 4]
```diff
@@ -185,6 +200,137 @@ public Response isInstanceStoppable(
...
+      inputFields.healthCheckConfig.remove(continueOnFailuresName);
+      inputFields.healthCheckConfig.remove(skipZKReadName);
+    }
+
+    inputFields.operations = MaintenanceManagementService
+        .getListFromJsonPayload(node.get(PerInstanceProperties.operation_list.name()));
+    inputFields.operationConfig = MaintenanceManagementService
+        .getMapFromJsonPayload(node.get(PerInstanceProperties.operation_config.name()));
+
+    LOG.debug("healthChecks: " + (inputFields.healthChecks == null ? "NULL"
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java:323]: Can we just use inputFields.toString()?

**Review by junkaixue — COMMENTED** [review]
Let's have tests for the APIs added

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java [hunk 5]
```diff
@@ -423,6 +422,23 @@ private StoppableCheck performCustomInstanceCheck(String clusterId, String insta
...
+      return Collections.emptyList();
+    }
+    return OBJECT_MAPPER.convertValue(jsonContent, List.class);
+  }
+
+  public static Map<String, String> getMapFromJsonPayload(JsonNode jsonContent) throws IOException {
+    Map<String, String> result = new HashMap<>();
+    if (jsonContent == null) {
+      return result;
+    }
```
Comment (narendly) [inline helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:436]: You could simplify this logic with the use of a ternary operator, which would let you save the work of instantiating an empty HashMap instance.

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/api/OperationAbstractClass.java [hunk 6]
```diff
@@ -26,7 +26,7 @@
 import org.apache.helix.rest.common.datamodel.RestSnapShot;
 
 
-interface OperationAbstractClass {
+public interface OperationAbstractClass {
```
Comment (narendly) [inline helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/api/OperationAbstractClass.java:29]: An interface with "Abstract" in its name seems like a code smell?

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java (hunk 5 above)
Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java]: TFTR. Good suggestion. Updated.

**Review by junkaixue — APPROVED** [review]

**Comment by xyuanlu** [comment]
This PR is ready to be checked in. Approved by @junkaixue  //Thanks!
Commit message:
Add rest API for take/free instance

**Pull request merged**